crates/uv-trampoline/src/bounce.rs

```diff
@@ -91,27 +109,82 @@ fn find_python_exe(is_gui: bool) -> CString {
                 break;
             }
         }
-        // Replace the filename (the last segment of the path) with "python.exe"
-        // Assumption: We are not in an encoding where a backslash byte can be part of a larger character.
-        let Some(last_backslash) = buffer.iter().rposition(|byte| *byte == b'\\') else {
-            eprintln!(
-                "Invalid current exe path (missing backslash): `{}`",
-                CString::from_vec_with_nul_unchecked(buffer)
-                    .to_string_lossy()
-                    .as_ref()
-            );
-            ExitProcess(1);
-        };
-        buffer.truncate(last_backslash + 1);
-        buffer.extend_from_slice(if is_gui {
-            b"pythonw.exe\0"
-        } else {
-            b"python.exe\0"
-        });
         CString::from_vec_with_nul_unchecked(buffer)
     }
 }
 
+/// The scripts are in the same directory as the Python interpreter, so we can find Python by getting the locations of
+/// the current .exe and replacing the filename with `python[w].exe`.
+fn find_python_exe(is_gui: bool, executable_name: &CStr) -> CString {
+    // Replace the filename (the last segment of the path) with "python.exe"
+    // Assumption: We are not in an encoding where a backslash byte can be part of a larger character.
+    let bytes = executable_name.to_bytes();
+    let Some(last_backslash) = bytes.iter().rposition(|byte| *byte == b'\\') else {
+        eprintln!(
+            "Invalid current exe path (missing backslash): `{}`",
+            &*executable_name.to_string_lossy()
+        );
+        unsafe {
+            ExitProcess(1);
+        }
+    };
+
+    let mut buffer = bytes[..last_backslash + 1].to_vec();
+    buffer.extend_from_slice(if is_gui {
+        b"pythonw.exe"
+    } else {
+        b"python.exe"
+    });
+    buffer.push(b'\0');
+
+    unsafe { CString::from_vec_with_nul_unchecked(buffer) }
```

Comment:
I'd drop the `unsafe` here too. (Although I do believe it is correct.) And replace it with [`CString::new(buffer).expect("no interior NUL bytes")`](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.new).

Reply:
The null terminator on line 138 should guarantee the null termination ;) 

I tried to keep the implementation consistent with the existing code that exclusively use the unchecked variation because of what's mentioned here

https://github.com/astral-sh/uv/blob/2586f655bbf650a9797c8f88b6d9066eefe7a3dc/crates/uv-trampoline/README.md#L75-L86

I don't mind reconsidering the decision to prefer checked versions over unchecked ones but we might want to do this separately because we should look at the generated artifact size etc.

Reply:
> I don't mind reconsidering the decision to prefer checked versions over unchecked ones but we might want to do this separately because we should look at the generated artifact size etc.

I reluctantly agree that this is probably the wise course of action for now.

Yeah, binary size is a primary concern for this particular library but I would like to eventually litigate whether it's a primary concern for us.